core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/WindowsShareCopy.java

```diff
@@ -0,0 +1,275 @@
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import com.google.common.io.ByteStreams;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import jcifs.smb.NtlmPasswordAuthentication;
+import jcifs.smb.SmbFile;
+import org.apache.avro.reflect.Nullable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link WindowsShareCopy} is a {@link Action} that will copy the data from Window share into HDFS directory.
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("WindowsShareCopy")
+@Description("Copies Files from Windows Share into HDFS Directory without intermediate store.")
+public class WindowsShareCopy extends Action {
+  private static final Logger LOG = LoggerFactory.getLogger(WindowsShareCopy.class);
+  private static final int MIN_BUFFER_SIZE = 4096;
+  private static final int MIN_NUM_THREADS = 1;
+  private WindowsShareCopyConfig config;
+
+  public WindowsShareCopy(WindowsShareCopyConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    StringBuilder sb = new StringBuilder("smb://");
+    sb.append(config.netBiosHostname);
+    sb.append("/");
+    sb.append(config.netBiosSharename);
+    sb.append("/");
+    if (config.sourceDirectory.startsWith("/")) {
+      if (config.sourceDirectory.length() > 1) {
+        sb.append(config.sourceDirectory.substring(1));
+      }
+    } else {
+      sb.append(config.sourceDirectory);
+    }
+    final String smbDirectory = sb.toString();
+
+    // Register the SMB File handler.
+    jcifs.Config.registerSmbURLHandler();
+    // Set Jcifs Log level to log debug information also
+    jcifs.Config.setProperty("jcifs.util.loglevel", "4");
+
+    // Authentication with NTLM and read the directory from the Windows Share.
+    final NtlmPasswordAuthentication auth = new NtlmPasswordAuthentication(config.netBiosDomainName,
+                                                                           config.netBiosUsername,
+                                                                           config.netBiosPassword);
+    SmbFile dir = new SmbFile(smbDirectory, auth);
+
+    // Determines the buffer size to be used during writing.
+    Configuration conf = new Configuration();
+    conf.setLong("io.file.buffer.size", config.bufferSize);
```

Comment:
I don't really understand why is this done and how its being used to control the buffer size. Can you explain ?